File: Number of islands BFS.java
// Time complexity 0(m*n) no of rows and columns
// space complexity 0(n) 
// code successfully executed in leetcode.
class Solution {
    public int numIslands(char[][] grid) {
        int m = grid.length;
        if(grid == null || m == 0) return 0;
        int n = grid[0].length;
        int count =0;
        int [][] dirs ={{0,1},{0,-1},{1,0},{-1,0}};
        //Queue<Integer []> q1 = new LinkedList<>();
        for(int i=0;i<m;i++){
            for(int j=0;j<n;j++){
                if(grid[i][j] == '1'){
                    count++;
                    Queue<int []> q1 = new LinkedList<>();
                    q1.add(new int []{i,j});
                    grid[i][j] ='0';
                    while(!q1.isEmpty()){
                        int curr [] = q1.poll();
                        for(int [] dir: dirs){
                            int r = dir[0] + curr[0];
                            int c = dir[1] + curr[1];
                 if(r >=0 && r<m && c>=0 && c<n && grid[r][c] == '1'){
                     q1.add(new int [] {r,c});
                     grid[r][c] = '0';                     
                 }
                        }
                    }   
                }  
            }
        }
        
       return count; 
    }
}

File: NumberOfIslands_BFS.py
# TC: O(M x N) since we will end up processing all the characters in the input array. 
# SC: O(min(M,N)) where M is the number of rows and N os the number of columns. 

class Solution:
    def numIslands(self, grid: List[List[str]]) -> int:
        if not grid or len(grid) == 0 or len(grid[0]) == 0:
            return 
        
        count = 0
        queue = collections.deque()
        rows = len(grid)
        cols = len(grid[0])
        dirs = [(0,1),(0,-1),(1,0),(-1,0)]
        for i in range(rows):
            for j in range(cols):
                if grid[i][j] == '1':
                    count += 1
                    grid[i][j] = '0'
                    queue.append((i,j))
                    while queue:
                        curr_x, curr_y = queue.popleft()
                        for x,y in dirs:
                            new_x, new_y = curr_x + x, curr_y + y
                            if new_x >= 0 and new_x < rows and new_y >= 0 and new_y < cols and grid[new_x][new_y] == '1':
                                queue.append((new_x, new_y))
                                grid[new_x][new_y] = '0'
        return count
